### ISSUE TITLE
Set up a proper pipeline for GHA

### ISSUE THREAD

**Issue by chadwhitacre** [getsentry/.github#22]
### Problem

We are like 1990s PHP developers right now. We have no development pipeline, no tests, we're copying and pasting files all over the place, we're editing live on the server. How can we bring normal software development practices to this strange new serverless GHA world?

### Solution

I want to store workflows in `.github` and deploy from there out to target repos, in phases (i.e., `onpremise`, then `sentry-docs`, then `sentry`, then ... ?). I envision an action here in `.github` to do the deploy, with `target_repo` as a parameter (probably required, with a special case of `all` for GA deploy ... or maybe repo groups are defined in a yaml or something).

That's deploy. CI? I want to make a PR against this `.github` repo—as I've done here—and have CI kick in that runs the tests, but here's the catch: they need to run in a cleanroom repo, we don't want test fixture to clutter up this repo. I made a `.github-ci` repo to use for that, but now I think we probably actually want to [create repos on the fly](https://docs.github.com/en/rest/reference/repos#create-an-organization-repository) so that multiple CI runs in parallel do not clobber each other. I envision an action here in `.github` to do that, wired up like normal CI in any other repo ([e.g.](https://github.com/getsentry/onpremise/blob/21.1.0/.github/workflows/test.yml)).

Okay! Deploy and CI! How about dev? I made a `.github-chadwhitacre` repo for myself that has `.github` as an upstream, so I can stay in sync and do my dev work off of the latest. The best way to develop GitHub Actions is by running them in a true GitHub Actions environment, that is to say, in the true actual cloud. ☁️  Trying to use [nektos/act](https://github.com/nektos/act) is going to be too heavy and drifty by comparison. 🚢  I made a helper script called [`ons`](https://gist.github.com/chadwhitacre/a6bf2a7e5fc2a6222a7bd53b9b7626c2) (for "on save" ... also considering `gah` as a play on **G**it**H**ub **A**ctions and the exclamation "Gah!!") that commits, pushes, and kicks off test actions in my dev repo. I've been running it manually but it could maybe be hooked up to run on save in one's editor if so desired. My workflow is:

1. Edit files locally.
2. Save.
3. `ons`
4. Go to browser tab with dev repo "Actions" open, refresh.
5. Click, click to current test run.
6. Inspect output.
7. Repeat.

Test actions are named `foo-test.yml` (à la `foo_test.go`).

Here's a wrinkle: I want to keep test yamls out of the main `.github` repo, just like we don't ship test code to prod in other software architectures. Actually, nevermind: `.github` is not the prod environment, it's the final step before prod. The action that deploys action yamls to the actual target repos will not ship test yamls. Done!

### Todo

- [ ] Write CI for `.github`.
- [ ] Write an action deploy action ([starter script](https://gist.github.com/chadwhitacre/76edf3b7392373a3b90bb4bfe4152960)).
- [ ] Clean up and publish `ons`.
- [ ] https://github.com/getsentry/.github/issues/30
- [x] https://github.com/getsentry/.github/issues/31

**Comment by BYK** [comment]
I like the general direction of this. We can use template repos feature to have the "base test repo" and keep generating repos from that. Not sure if GitHub would be happy about it but until they start complaining, I like the idea.

I'm a bit concerned about the long iteration cycle as the `ons` script will still use a personalized repo and runs on the cloud (so friction and latency) but maybe this is not too bad.

Finally, we should split this into phases (CI and deploy at least) and spec out a bit clearer before diving in.

**Comment by chadwhitacre** [comment]
Capturing some of our discussion about testing here as well: rather than asserting on log output, we should assert on resulting GitHub object state (e.g., issue labels in the case of validation, which is the first bot we have [tests](https://github.com/getsentry/.github/blob/61a392d38bfa7bd4a8b9b4cafd28b2ad5cb8bb37/.github/workflows/validate-new-issue-test.yml) for). Another pattern to adopt is to unit test in a regular language runtime, and then only do integration tests in GHA (as suggested in "[How to Test Serverless Applications](https://www.serverless.com/blog/how-test-serverless-applications)").

**Comment by chadwhitacre** [comment]
GH workflow and action validator for pre-commit, h/t @armenzg:

https://github.com/sirosen/check-jsonschema

**Comment by chadwhitacre** [comment]
https://github.com/fregante/ghat

**Comment by chadwhitacre** [comment]
Switching focus to eng-pipes app, not investing further in GHA infra.
